stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -547,54 +464,70 @@ struct adj_jac_vari : public vari {
  * So implementing f(x1, x2) and the products above are all that is required
  * mathematically to implement reverse-mode autodiff for a function.
  *
- * adj_jac_apply takes as a template argument a functor F that supplies the
- * non-static member functions (leaving exact template arguments off):
- *
- * (required) Eigen::VectorXd operator()(const std::array<bool, size>&
- * needs_adj, const T1& x1..., const T2& x2, ...)
- *
- * where there can be any number of input arguments x1, x2, ... and T1, T2,
- * ... can be either doubles or any Eigen::Matrix type with double scalar
- * values. needs_adj is an array of size equal to the number of input
- * arguments indicating whether or not the adjoint^T Jacobian product must be
- * computed for each input argument. This argument is passed to operator() so
- * that any unnecessary preparatory calculations for multiply_adjoint_jacobian
- * can be avoided if possible.
+ * adj_jac_apply takes as a template argument a functor F of the form
+ * ```
+ * template <typename T1, typename T2> // Any number of templates
+ * class my_functor {
+ *  // Helper type trait that is an eigen map for containers or scalar
+ *  // for scalars
+ *  adj_arg_t<T1> A_;
+ *  adj_arg_t<T2> B_;
+ *  // etc.
  *
- * (required) std::tuple<T1, T2, ...> multiply_adjoint_jacobian(const
- * std::array<bool, size>& needs_adj, const Eigen::VectorXd& adj)
+ *  // Constructor to allocate memory
+ *  // Must accept same types as T1 and T2
+ *  my_functor(const T1& A, const T2& B) :
+ *    A_(setup_adj_arg<T1>(A.rows(), A.cols()))
+ *    B_(setup_adj_arg<T2>(B.rows(), B.cols())) {}
+ *  // Member function to calculate forward pass values
+ *  // Must accept same types as the result of calling value_of() on
+ *  //  T1 and T2 types
+ *  template <typename S1, typename S2>
+ *  auto forward_pass(S1&& A, S2&& B) {
+ *    // Calculate forward pass value
+ *  }
+ *  // Calculate adjoint values in reverse pass
+ *  // Argument type must be the same as result of forward_pass() and
```

Comment:
The argument will be the adjoints of the output from the forward pass (and have the same type and shape as this output).

The `reverse_pass` function is expected to compute the adjoints of the inputs to the forward pass given the adjoints of the output from the forward pass.